pennylane/gradients/jvp.py

```diff
@@ -88,47 +78,123 @@ def compute_jvp_single(tangent, jac):
+             - ``jac`` shape
+             - Comment
+           * - ``(k,)`` or ``[(),..,()]`` or ``((),..,())``
+             - ``((),..,())`` (length ``k``)
+             - scalar return, ``k`` scalar parameters
+           * - ``(k,)`` or ``[(),..,()]`` or ``((),..,())``
+             - ``((r_1,..,r_R),..,(r_1,..,r_R))``
+             - tensor return, ``k`` scalar parameters
+           * - ``[(l_1,..,l_{L_1}),..,(l_1,..,l_{L_k})]``
+             - ``((l_1,..,l_{L_1}),..,(l_1,..,l_{L_k}))``
+             - scalar return, ``k`` tensor parameters
+           * - ``[(l_1,..,l_{L_1}),..,(l_1,..,l_{L_k})]``
+             - ``((r_1,..,r_R, l_1,..,l_{L_1}),..,(r_1,..,r_R, l_1,..,l_{L_k}))``
+             - tensor return, ``k`` tensor parameters
+
+        Note that the return type dimensions always remain the same, whereas the
+        dimensions of the tensor parameters may differ.
+
+        In this scenario, another case separation is used: If any of the parameters is a
+        tensor (i.e. not a scalar), all tangent entries are reshaped into one-dimensional
+        tensors with shapes ``(prod_{l_i},)`` and then stacked into one one-dimensional tensor.
+        If there are no tensor parameters, the tangent is just stacked and reshaped.
+        The Jacobians are reshaped to have the dimensions ``(r_1, ... r_R, prod_{l_i})``
+        and then are concatenated along their last (potentially mismatching) axis.
+        This is followed by a tensordot contraction over the axes of size
+        :math:`\sum_i` ``prod_{l_i}``.
+
     """
     if jac is None:
         return None
-
-    tangent = qml.math.stack(tangent)
-    jac = _convert(jac, tangent)
-
-    # Single param
-    if not isinstance(jac, tuple):
+    single_param = not isinstance(jac, tuple)
+    if (single_param and jac.shape == (0,)) or (not single_param and len(jac) == 0):
         # No trainable parameters
-        if jac.shape == (0,):
-            res = qml.math.zeros((1, 0))
-            return res
-
-        tangent = qml.math.reshape(tangent, (1,))
-
-        # No dimension e.g. expval
-        if jac.shape == ():
-            jac = qml.math.reshape(jac, (1,))
-
-        # With dimension e.g. probs
-        else:
-            jac = qml.math.reshape(jac, (1, -1))
-        res = qml.math.tensordot(jac, tangent, [[0], [0]])
-    # Multiple params
+        return qml.math.zeros((1, 0))
+
+    if single_param:
+        tangent = qml.math.stack(tangent)
+        L_1 = len(tangent.shape[1:])
+        tangent = qml.math.reshape(tangent, (-1,))
+        prod_l = tangent.shape[0]
+        shape = jac.shape
+        new_shape = shape[: len(shape) - L_1] + (prod_l,)
+        jac = qml.math.cast(qml.math.convert_like(jac, tangent), tangent.dtype)
+        jac = qml.math.reshape(jac, new_shape)
+        return qml.math.tensordot(jac, tangent, [[-1], [0]])
+
+    L = [t.ndim for t in tangent]
+    if isinstance(tangent, (tuple, list)) and any(t.ndim > 0 for t in tangent):
```

Comment:
Ah, but sometimes it is a `tensor_like`, isn't it? :thinking:

Reply:
I checked and this actually happens. The additional `isinstance` check allows us to avoid iterating over a one-dimensional tangent, e.g. a `torch.tensor` or so, and checking each entry to be non-scalar. However, if `tangent` is such a `tensor`, we already know it is not ragged and we can skip that iterative check.

Reply:
Ok thanks!